nifi-nar-bundles/nifi-standard-services/nifi-oauth2-provider-bundle/nifi-oauth2-provider-service/src/main/java/org/apache/nifi/oauth2/StandardOauth2AccessTokenProvider.java

```diff
@@ -0,0 +1,293 @@
+
+    private volatile AccessToken accessDetails;
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @OnEnabled
+    public void onEnabled(ConfigurationContext context) {
+        authorizationServerUrl = context.getProperty(AUTHORIZATION_SERVER_URL).evaluateAttributeExpressions().getValue();
+
+        httpClient = createHttpClient(context);
+
+        grantType = context.getProperty(GRANT_TYPE).getValue();
+        username = context.getProperty(USERNAME).evaluateAttributeExpressions().getValue();
+        password = context.getProperty(PASSWORD).getValue();
+        clientId = context.getProperty(CLIENT_ID).evaluateAttributeExpressions().getValue();
+        clientSecret = context.getProperty(CLIENT_SECRET).getValue();
+    }
+
+    @OnDisabled
+    public void onDisabled() {
+        accessDetails = null;
+    }
+
+    @Override
+    protected Collection<ValidationResult> customValidate(ValidationContext validationContext) {
+        final List<ValidationResult> validationResults = new ArrayList<>(super.customValidate(validationContext));
+
+        if (
+            validationContext.getProperty(GRANT_TYPE).getValue().equals(CLIENT_CREDENTIALS_GRANT_TYPE.getValue())
+                && !validationContext.getProperty(CLIENT_ID).isSet()
+        ) {
+            validationResults.add(new ValidationResult.Builder().subject(CLIENT_ID.getDisplayName())
+                .valid(false)
+                .explanation(String.format(
+                    "When '%s' is set to '%s', '%s' is required",
+                    GRANT_TYPE.getDisplayName(),
+                    CLIENT_CREDENTIALS_GRANT_TYPE.getDisplayName(),
+                    CLIENT_ID.getDisplayName())
+                )
+                .build());
+        }
+
+        return validationResults;
+    }
+
+    protected OkHttpClient createHttpClient(ConfigurationContext context) {
+        OkHttpClient.Builder clientBuilder = new OkHttpClient.Builder();
+
+        SSLContextService sslService = context.getProperty(SSL_CONTEXT).asControllerService(SSLContextService.class);
+        if (sslService != null) {
+            final X509TrustManager trustManager = sslService.createTrustManager();
+            SSLContext sslContext = sslService.createContext();
+            clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), trustManager);
+        }
+
+        return clientBuilder.build();
+    }
+
+    @Override
+    public AccessToken getAccessDetails() {
+        if (this.accessDetails == null) {
+            acquireAccessDetails();
+        } else if (this.accessDetails.isExpired()) {
+            try {
+                refreshAccessDetails();
+            } catch (Exception e) {
+                getLogger().info("Couldn't refresh access token, probably refresh token has expired." +
+                    " Getting new token.");
+                acquireAccessDetails();
+            }
```

Comment:
What about the case of an `IOException` thrown from `refreshAccessDetails()`? Whether wrapped in a `ProcessException` or `UncheckedIOException`, this will hit this catch block. In that case, the failure is not due to an expired refresh token, but a network issue like a SocketTimeoutException.

At minimum, the log message is not indicative of that type of failure. It also seems like making another call to the service with `acquireAuthorizationDetails()` is unlikely to succeed.  One option is to catch and throw `IOExceptions`.

```suggestion
            } catch (final IOException ioe) {
                throw new UncheckedIOException("Refresh Access Token Failed", ioe);
            } catch (final Exception e) {
                getLogger().info("Refresh Access Token Failed [{}]: New Token request started", e.getMessage());
                acquireAccessDetails();
            }
```

Reply:
We might save a round to the access server with this approach but that's about the only advantage. However a simple retry (in this case, try to get a whole new access token) might even provide a solution for a momentary connection issue. Overall I still think the current implementation is more optimal.

As for troubleshooting I stand by my previous opinion: I tested all error scenarios and in all cases I see the issue properly reported. The information provided is clear and lacks no detail.
Unless presented a concrete detailed and reproducible example that would prove otherwise I consider this level of error handling sufficient.

Reply:
If the initial request for a token fails, it is important to know the reason, and OAuth2 responses provide specific details for the failure. Another problem with the current implementation is that the exception details are not logged.

@MikeThomsen can probably provide more detail on the reason that the current implementation separates the initial token request and the refresh token request. Although there is more work on the caller, these are two different concepts that can fail for different reasons.

Although the general functionality would be an improvement to InvokeHTTP, I am -1 on the current approach. If token refresh remains handled implicitly in the Controller Service, the failure conditions need to be handled more clearly.

Reply:
> If the initial request for a token fails, it is important to know the reason, and OAuth2 responses provide specific details for the failure.

I agree. I tested this scenario and - as I said previously: I see the issue properly reported. The information provided is clear and lacks no detail.

> Another problem with the current implementation is that the exception details are not logged.

Okay, I'll add the exception to the logger call.

If you still think you can present a concrete example where you can show that a certain issue can't be investigated efficiently or a case is not handled properly please give me the details how to reproduce it and I'd be very happy to take a look.